src/processor.cpp

```diff
@@ -0,0 +1,737 @@
+// enabled will be used. If there's still a tie the one that appears later in the list will be
+// used. (i.e. the order in the version list is significant in this case).
+
+// Features that are not recognized will be passed to LLVM directly during codegen
+// but ignored otherwise.
+
+// Two special features are supported:
+// 1. `clone_all`
+//
+//     This forces the target to have all functions in sysimg cloned.
+//     When used in negative form (i.e. `-clone_all`), this disables full clone that's
+//     enabled by default for certain targets.
+//
+// 2. `base([0-9]*)`
+//
+//     This specifies the (0-based) base target index. The base target is the target
+//     that the current target is based on, i.e. the functions that are not being cloned
+//     will use the version in the base target. This option causes the base target to be
+//     fully cloned (as if `clone_all` is specified for it) if it is not the default target (0).
+//     The index can only be smaller than the current index.
+
+namespace {
+
+// Helper functions to test/set feature bits
+
+template<typename T1, typename T2, typename T3>
+static inline bool test_bits(T1 v, T2 mask, T3 test)
+{
+    return T3(v & mask) == test;
+}
+
+template<typename T1, typename T2>
+static inline bool test_bits(T1 v, T2 mask)
+{
+    return test_bits(v, mask, mask);
+}
+
+template<typename T1, typename T2>
+static inline bool test_nbit(const T1 &bits, T2 _bitidx)
+{
+    auto bitidx = static_cast<uint32_t>(_bitidx);
+    auto u32idx = bitidx / 32;
+    auto bit = bitidx % 32;
+    return (bits[u32idx] & (1 << bit)) != 0;
+}
+
+template<typename T>
+static inline void unset_bits(T &bits)
+{
+    (void)bits;
+}
+
+template<typename T, typename T1, typename... Rest>
+static inline void unset_bits(T &bits, T1 _bitidx, Rest... rest)
+{
+    auto bitidx = static_cast<uint32_t>(_bitidx);
+    auto u32idx = bitidx / 32;
+    auto bit = bitidx % 32;
+    bits[u32idx] = bits[u32idx] & ~uint32_t(1 << bit);
+    unset_bits(bits, rest...);
+}
+
+template<typename T, typename T1>
+static inline void set_bit(T &bits, T1 _bitidx, bool val)
+{
+    auto bitidx = static_cast<uint32_t>(_bitidx);
+    auto u32idx = bitidx / 32;
+    auto bit = bitidx % 32;
+    if (val) {
+        bits[u32idx] = bits[u32idx] | uint32_t(1 << bit);
+    }
+    else {
+        bits[u32idx] = bits[u32idx] & ~uint32_t(1 << bit);
+    }
+}
+
+// Helper functions to create feature masks
+
+// This can be `std::array<uint32_t,n>` on C++14
+template<size_t n>
+struct FeatureList {
+    uint32_t eles[n];
+    uint32_t &operator[](size_t pos)
+    {
+        return eles[pos];
+    }
+    constexpr const uint32_t &operator[](size_t pos) const
+    {
+        return eles[pos];
+    }
+    inline int nbits() const
+    {
+        int cnt = 0;
+        for (size_t i = 0; i < n; i++) {
+#ifdef __GNUC__
+            cnt += __builtin_popcount(eles[i]);
```

Comment:
use the llvm names (e.g. `llvm::countPopulation(eles[i])` in this case), from `#include <llvm/Support/MathExtras.h>`